File: src/org/simbrain/custom_sims/simulations/behaviorism/OperantWithEnvironment.java
package org.simbrain.custom_sims.simulations.behaviorism;

import org.simbrain.custom_sims.RegisteredSimulation;
import org.simbrain.custom_sims.helper_classes.ControlPanel;
import org.simbrain.custom_sims.helper_classes.NetBuilder;
import org.simbrain.custom_sims.helper_classes.OdorWorldBuilder;
import org.simbrain.network.core.Network;
import org.simbrain.network.core.NetworkUpdateAction;
import org.simbrain.network.core.Neuron;
import org.simbrain.network.core.Synapse;
import org.simbrain.network.groups.NeuronGroup;
import org.simbrain.network.layouts.LineLayout;
import org.simbrain.network.neuron_update_rules.BinaryRule;
import org.simbrain.util.math.SimbrainMath;
import org.simbrain.workspace.gui.SimbrainDesktop;
import org.simbrain.world.odorworld.entities.OdorWorldEntity;
import org.simbrain.world.odorworld.entities.RotatingEntity;
import org.simbrain.world.odorworld.sensors.SmellSensor;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Simulation to demonstrate classical and operant conditioning.
 * Discriminative case
 *
 * @author Tim Meyer
 * @author Jeff Yoshimi
 */
public class OperantWithEnvironment extends RegisteredSimulation {

    // Network
    NetBuilder netBuilder;
    Network network;
    ControlPanel panel;
    NeuronGroup behaviorNet;
    NeuronGroup stimulusNet;
    Neuron rewardNeuron, punishNeuron;
    Map<Neuron, String> nodeToLabel = new HashMap();
    final int numNeurons = 3;
    double[] firingProbabilities = new double[numNeurons];
    int winningNode;

    // World
    OdorWorldBuilder world;
    RotatingEntity mouse;
    OdorWorldEntity cheese, flower, fish;

    public OperantWithEnvironment() {
        super();
    }

    public OperantWithEnvironment(SimbrainDesktop desktop) {
        super(desktop);
    }

    @Override
    public void run() {

        // Clear workspace
        sim.getWorkspace().clearWorkspace();
        netBuilder = sim.addNetwork(155,9,575,500, "Brain");
        network = netBuilder.getNetwork();

        // Behavioral nodes
        behaviorNet = netBuilder.addNeuronGroup(-9.25, 95.93, numNeurons);
        ((LineLayout) behaviorNet.getLayout()).setSpacing(100);
        behaviorNet.applyLayout();
        behaviorNet.setLabel("Behaviors");

        // Stimulus nodes
        stimulusNet = netBuilder.addNeuronGroup(-9.25, 295.93, numNeurons);
        ((LineLayout) stimulusNet.getLayout()).setSpacing(100);
        stimulusNet.applyLayout();
        stimulusNet.setClamped(true);
        stimulusNet.setLabel("Stimuli");
        stimulusNet.setIncrement(1);

        // Reward and punish nodes
        rewardNeuron = netBuilder.addNeuron((int)stimulusNet.getMaxX() + 100,
            (int) stimulusNet.getCenterY());
        rewardNeuron.setLabel("Food Pellet");
        punishNeuron = netBuilder.addNeuron((int) rewardNeuron.getX() + 100,
            (int) stimulusNet.getCenterY());
        punishNeuron.setLabel("Shock");

        // Set base text for behavior labels
        nodeToLabel.put(behaviorNet.getNeuron(0), "Wiggle");
        nodeToLabel.put(behaviorNet.getNeuron(1), "Explore");
        nodeToLabel.put(behaviorNet.getNeuron(2), "Spin");

        // Set stimulus labels
        stimulusNet.getNeuron(0).setLabel("Cheese");
        stimulusNet.getNeuron(1).setLabel("Flower");
        stimulusNet.getNeuron(2).setLabel("Bell");

        // Use aux values to store "intrinsic" firing probabilities for behaviors
        behaviorNet.getNeuron(0).setAuxValue(.33); // Node 0
        behaviorNet.getNeuron(1).setAuxValue(.33); // Node 1
        behaviorNet.getNeuron(2).setAuxValue(.33); // Node 2

        // Initialize behaviorism labels
        updateNodeLabels();

        // Connect the layers together
        List<Synapse> syns = netBuilder.connectAllToAll(stimulusNet, behaviorNet);
        for(Synapse s : syns) {
            s.setStrength(0);
        }
        network.fireSynapsesUpdated();

        // Create the odor world
        world = sim.addOdorWorld(730,7,315,383, "Three Objects");
        world.getWorld().setObjectsBlockMovement(false);
        mouse = world.addAgent(120, 245, "Mouse");
        mouse.setHeading(90);

        // Set up world
        cheese = world.addEntity(27, 20, "Swiss.gif",
            new double[] { 1, 0, 0 });
        cheese.getSmellSource().setDispersion(65);
        flower = world.addEntity(79, 20, "Pansy.gif",
            new double[] { 0, 1, 0 });
        cheese.getSmellSource().setDispersion(65);
        fish = world.addEntity(125, 20, "Bell.gif",
            new double[] { 0, 0, 1 });
        cheese.getSmellSource().setDispersion(65);

        // Couple agent to network
        sim.couple((SmellSensor) mouse.getSensor("Smell-Center"), 0,
            stimulusNet.getNeuron(0));
        sim.couple((SmellSensor) mouse.getSensor("Smell-Center"), 1,
            stimulusNet.getNeuron(1));
        sim.couple((SmellSensor) mouse.getSensor("Smell-Center"), 2,
            stimulusNet.getNeuron(2));

        // Add custom network update action
        network.getUpdateManager().addAction(0,new NetworkUpdateAction() {

            @Override
            public void invoke() {
                updateNetwork();
                updateBehaviors();
            }

            @Override
            public String getDescription() {
                return "Custom behaviorism update";
            }

            @Override
            public String getLongDescription() {
                return "Custom behaviorism update";
            }
        });

        setUpControlPanel();

    }

    /**
     * Update behavior of odor world agent based on which node is active.
     * Assumes behaviors partitioned into increments of (currently) 100 time steps
     */
    private void updateBehaviors() {

        int loopTime = sim.getWorkspace().getTime() % 10;

        // Node 0: Wiggle
        if(winningNode == 0) {
            if (loopTime < 5) {
                mouse.setHeading(mouse.getHeading() + 5);
            } else {
                mouse.setHeading(mouse.getHeading() - 5);
            }
        }
        // Node 1: Explore
        else if (winningNode == 1) {
            if (Math.random() < .2) {
                mouse.setHeading(mouse.getHeading() + Math.random()*20-10);
            }
            mouse.goStraight(2.5);
        }
        // Node 2: Spin
        else {
            mouse.setHeading(mouse.getHeading()+20);
        }
    }

    private void updateNetwork() {

        // Update firing probabilities
        for (int i = 0; i < behaviorNet.size(); i++) {
            Neuron n = behaviorNet.getNeuron(i);
            firingProbabilities[i] = n.getWeightedInputs() + n.getAuxValue();
        }
        firingProbabilities = SimbrainMath.normalizeVec(firingProbabilities);
        //System.out.println(Arrays.toString(firingProbabilities));

        // Select "winning" neuron based on its probability
        double random = Math.random();
        if(random < firingProbabilities[0]){
            setWinningNode(0);
        } else if(random < firingProbabilities[0] + firingProbabilities[1]) {
            setWinningNode(1);
        } else{
            setWinningNode(2);
        }

    }

    private void setWinningNode(int nodeIndex) {
        winningNode = nodeIndex;
        for (int i = 0; i < behaviorNet.size(); i++) {
            if (i == nodeIndex) {
                behaviorNet.getNeuron(i).setInputValue(1);
            } else {
                behaviorNet.getNeuron(i).setInputValue(0);
            }
        }
    }

    private void setUpControlPanel() {

        panel = ControlPanel.makePanel(sim, "Control Panel", 5, 10);

        panel.addButton("Reward", () -> {
            learn(1);
            rewardNeuron.setInputValue(1);
            punishNeuron.forceSetActivation(0);
            sim.iterate();
        });

        panel.addButton("Punish", () -> {
            learn(-1);
            rewardNeuron.forceSetActivation(0);
            punishNeuron.setInputValue(1);
            sim.iterate();
        });

        panel.addButton("Do nothing", () -> {
            rewardNeuron.forceSetActivation(0);
            punishNeuron.setInputValue(0);
            sim.iterate();
        });


    }

    private void learn(double valence) {

        double rewardLearningRate = .1;
        double punishLearningRate = .1;

        if(valence > 0) {
            valence *= rewardLearningRate;
        } else {
            valence *= punishLearningRate;
        }

        for(Neuron tar : behaviorNet.getNeuronList()) {

            // The "winning" node
            if(tar.getActivation() > 0){

                // Update intrinsic probability
                double p = tar.getAuxValue();
                tar.setAuxValue(Math.max(p + valence * p, 0));

                // Update weight on active node
                for(Neuron src : stimulusNet.getNeuronList()) {
                    if (src.getActivation() > 0) {
                        Synapse s = Network.getSynapse(src,tar);
                        s.setStrength(Math.max(s.getStrength() + valence, 0));
                    }
                }
            }
        }
        normIntrinsicProbabilities();
        updateNodeLabels();
        sim.iterate();
    }

    private void normIntrinsicProbabilities() {
        double totalMass = 0;
        for(Neuron n : behaviorNet.getNeuronList()) {
            totalMass += n.getAuxValue();
        }
        for(Neuron n : behaviorNet.getNeuronList()) {
            n.setAuxValue(n.getAuxValue()/totalMass);
        }
    }

    private void updateNodeLabels() {
        for(Neuron n : behaviorNet.getNeuronList()) {
            n.setLabel(nodeToLabel.get(n) + ": "
                + SimbrainMath.roundDouble(n.getAuxValue(), 2));
        }
    }

    @Override
    public String getName() {
        return "Behaviorism: Operant with Mouse";
    }

    @Override
    public OperantWithEnvironment instantiate(SimbrainDesktop desktop) { return new OperantWithEnvironment(desktop);
    }

}

File: src/main/kotlin/org/simbrain/custom_sims/simulations/behaviorism/OperantWithEnvironment.kt
package org.simbrain.custom_sims.simulations.behaviorism

import org.simbrain.custom_sims.*
import org.simbrain.network.NetworkComponent
import org.simbrain.network.core.*
import org.simbrain.network.desktop.NetworkDesktopComponent
import org.simbrain.network.layouts.LineLayout
import org.simbrain.network.neurongroups.NeuronGroup
import org.simbrain.network.neurongroups.getWinner
import org.simbrain.util.*
import org.simbrain.util.piccolo.TileMap
import org.simbrain.workspace.Workspace
import org.simbrain.workspace.gui.SimbrainDesktop
import org.simbrain.world.odorworld.OdorWorldComponent
import org.simbrain.world.odorworld.OdorWorldDesktopComponent
import org.simbrain.world.odorworld.entities.EntityType
import kotlin.math.max
import kotlin.random.Random


val operantWithEnvironment = newSim("operant_with_environment") {
    workspace.clearWorkspace()

    val networkComponent = addNetworkComponent("Brain")
    val network = networkComponent.network

    val numNeurons = 3

    val behaviorNet = network.addNeuronGroup(numNeurons, location = point(-9.25, 95.93)).apply {
        layout = LineLayout(100.0, LineLayout.LineOrientation.HORIZONTAL)
        label = "Behaviors"
        neuronList.labels = listOf("Wiggle: ", "Explore: ", "Spin: ")
        neuronList.forEach { it.auxValue = .33 }
        applyLayout()
    }

    val stimulusNet = network.addNeuronGroup(numNeurons, location = point(-9.25, 295.93)).apply {
        layout = LineLayout(100.0, LineLayout.LineOrientation.HORIZONTAL)
        isClamped = true
        label = "Stimuli"
        setIncrement(1.0)
        neuronList.labels = listOf("Candle", "Flower", "Bell")
        applyLayout()
    }

    val rewardNeuron = network.addNeuron {
        location = point(stimulusNet.maxX + 100, stimulusNet.locationY)
        label = "Food Pellet"
        clamped = true
    }

    val punishNeuron = network.addNeuron {
        location = point(rewardNeuron.x + 100, rewardNeuron.locationY)
        label = "Shock"
        clamped = true
    }

    val syns = with(network) {
        connectAllToAll(stimulusNet, behaviorNet)
    }.onEach { it.strength = 0.0 }

    val odorWorldComponent = addOdorWorldComponent("Three Objects")
    val odorWorld = odorWorldComponent.world.apply {
        tileMap = TileMap(12, 12)
        isObjectsBlockMovement = false
        isUseCameraCentering = false
    }

    val mouse = odorWorld.addEntity(120, 245, EntityType.Mouse).apply {
        heading = 90.0
    }

    val candle = odorWorld.addEntity(27, 50, EntityType.Candle)
    val flower = odorWorld.addEntity(79, 50, EntityType.Pansy)
    val bell = odorWorld.addEntity(125, 50, EntityType.Bell)
    odorWorld.addEntity(candle)
    odorWorld.addEntity(flower)
    odorWorld.addEntity(bell)

    val cndleSensor = mouse.addObjectSensor(EntityType.Candle, 50.0, 0.0, 65.0)
    val flowerSensor = mouse.addObjectSensor(EntityType.Pansy, 50.0, 0.0, 65.0)
    val bellSensor = mouse.addObjectSensor(EntityType.Bell, 50.0, 0.0, 65.0)

    with(couplingManager) {
        val (n1, n2, n3) = stimulusNet.neuronList
        cndleSensor couple n1
        flowerSensor couple n2
        bellSensor couple n3
    }

    updateBehaviorNetNeuronLabels(behaviorNet)

    withGui {
        place(networkComponent, 155, 9, 575, 500)
        (getDesktopComponent(networkComponent) as NetworkDesktopComponent)
            .networkPanel.selectionManager.clear()

        place(odorWorldComponent, 730, 7, 315, 383)
        (getDesktopComponent(odorWorldComponent) as OdorWorldDesktopComponent).worldPanel.scalingFactor = .5
    }

    setupOperantWithEnvironmentWorkspace(workspace)

    addSidebarInfo(
        """
        
        # Operant with Environment
        
        This simulation demonstrates operant conditioning in an environment where an agent learns to associate stimuli with behaviors through reward and punishment. The agent has three
        basic behaviors (`Wiggle`, `Explore`, `Spin`) that it performs with certain probabilities. These behaviors can be reinforced or discouraged based on environmental feedback.
        
        # Simulation Details
        
        In this simulation, an agent (mouse) can sense three different objects: a `Candle`, `Flower`, and `Bell`.  
        The agent can perform three behaviors:
        
        - Wiggle: oscillates left and right  
        - Explore: moves around randomly  
        - Spin: rotates in place  
        
        The agent's behavior is controlled by a neural network with the following elements:
        
        - Stimulus neurons detect objects in the environment  
        - Behavior neurons control which action the agent performs  
        - Intrinsic probabilities (shown as labels on behavior neurons) determine baseline or spontaneous behavior tendencies  
        - Connections between stimuli and behaviors can be strengthened or weakened through learning. These connections determine behaviors that are conditional on the presence of 
        stimuli.  
        
        When the agent is near an object and performs a behavior, you can reward or punish that stimulus–response pairing. This increases or decreases the likelihood of that behavior
        occurring in that context.
        
        # What to Do
        
        1. Run the simulation and observe the agent's initial random behavior.  
        
        2. Train the agent to perform a behavior spontaneously. This is like the simple operant conditioning simulation.  
        
        3. Add a conditional behavior. Wait for the agent to approach an object (candle, flower, or bell). You will see the corresponding stimulus neuron activate.  
        
        4. Observe which behavior the agent performs near the object.  
        
        5. Provide feedback:  
           - Click `Reward` to encourage the stimulus–behavior pairing  
           - Click `Punish` to discourage it  
           - Click `Do nothing` to let the agent continue without feedback  
        
        6. Repeat the process. Over time, the agent will learn to associate certain stimuli with rewarded behaviors.  
        
        7. Monitor the changes:  
           - Watch how the connection weights between stimuli and behaviors change  
           - Notice how the agent's behavior becomes more predictable near certain objects  
           - Observe how intrinsic behavior probabilities (shown on the behavior neurons) adjust when no stimuli are present  
        
        At this point you can demonstrate the idea of a discriminative stimulus (also called a controlling stimulus). This is a stimulus that, after training, increases the probability
        of an operant behavior. It signals the relationship between a behavior and a reinforcer. The behavior can then be said to be under the control of that stimulus. For example, 
        pressing a lever may only produce food when a light is on.  
        
        In this simulation, you might first train the agent to wiggle spontaneously. Then you can transfer control of this behavior to the candle, so that wiggling only occurs when the
        candle is present. You could punish wiggling when the candle is absent, and train the agent to do something else spontaneously instead. In that case, wiggling is said to be under 
        the control of the candle, which is now the discriminative stimulus.
        
        # References
        
        Skinner, B. F. (1953). [_Science and human behavior_](https://www.bfskinner.org/newtestsite/wp-content/uploads/2014/02/ScienceHumanBehavior.pdf). Macmillan.

        # Credits
        
        Tim Meyer
        
        [Jeff Yoshimi](https://jeffyoshimi.net/index.html)
                
        """.trimIndent()
    )

}.registerReopenFunction { workspace -> setupOperantWithEnvironmentWorkspace(workspace) }

fun updateBehaviorNetNeuronLabels(behaviorNet: NeuronGroup) {
    behaviorNet.neuronList.forEach {
        it.label = it.label?.replace(Regex(":.+"), ": ${it.auxValue.format(2)}")
    }
}

suspend fun SimulationScope.setupOperantWithEnvironmentWorkspace(workspace: Workspace) {

    val random = Random(Random.nextLong())

    val network = workspace.componentList.filterIsInstance<NetworkComponent>().first().network
    val behaviorNet = network.getModelByLabel<NeuronGroup>("Behaviors")
    val stimulusNet = network.getModelByLabel<NeuronGroup>("Stimuli")
    val rewardNeuron = network.getModelByLabel<Neuron>("Food Pellet")
    val punishNeuron = network.getModelByLabel<Neuron>("Shock")

    val odorWorldComponent = workspace.componentList.filterIsInstance<OdorWorldComponent>().first()
    val odorWorld = odorWorldComponent.world
    val mouse = odorWorld.entityList.first { it.entityType == EntityType.Mouse }

    network.updateManager.clear()
    network.updateManager.addAction(updateAction("Custom behaviorism update") {

        var winningNode = 0

        /**
         * Update actual firing probabilities, which combine intrinsic probabilities with weighted inputs
         */
        suspend fun updateNetwork() {
            val firingProbabilityCDF = behaviorNet.neuronList
                .map { it.weightedInputs + it.auxValue }
                .let { if (it.any { v -> v < 0 }) it.minMaxNormalize() else it }
                .normalize()
                .runningReduce(Double::plus)

            val selection = random.nextDouble()

            // Select "winning" neuron based on its probability
            winningNode = firingProbabilityCDF.indexOfFirst {
                selection < it
            }

            // Set neuron activation based on winning node
            behaviorNet.neuronList.forEachIndexed { index, neuron -> neuron.addInputValue(if (index == winningNode) 1.0 else 0.0) }

            network.bufferedUpdate()
        }

        /**
         * Update behavior of odor world agent based on which node is active.
         * Assumes behaviors partitioned into increments of (currently) 100 time steps
         */
        fun updateBehaviors() {
            val loopTime = workspace.time % 10

            when (winningNode) {
                0 -> {
                    mouse.heading += if (loopTime < 5) 5 else -5
                }
                1 -> {
                    if (random.nextDouble() < 0.2) {
                        mouse.heading += random.nextDouble(-10.0, 10.0)
                    }
                    mouse.speed = 2.5
                }
                2 -> {
                    mouse.heading += 20
                }
            }
        }

        updateBehaviorNetNeuronLabels(stimulusNet)
        updateNetwork()
        updateBehaviors()
    })

    withGui {
        createControlPanel("Control Panel", 5, 10) {

            fun normIntrinsicProbabilities() {
                val totalMass = behaviorNet.neuronList.sumOf { it.auxValue }
                behaviorNet.neuronList.forEach { it.auxValue /= totalMass }
            }

            fun learn(initValence: Double) {
                val rewardLearningRate = .1
                val punishLearningRate = .1

                val valence = initValence * if (initValence > 0) {
                    rewardLearningRate
                } else {
                    punishLearningRate
                }

                val totalActivation = stimulusNet.neuronList.sumOf { it.activation }
                val winner = getWinner(behaviorNet.neuronList, true)

                // If there are inputs, update weights
                if (totalActivation > .1) {
                    val src = getWinner(stimulusNet.neuronList, true)
                    val s_r = getSynapse(src, winner) ?: throw IllegalStateException("Synapse not found")
                    // Strengthen or weaken active S-R Pair
                    s_r.strength += valence
                } else {
                    // Else update intrinsic probability
                    val p = winner.auxValue
                    winner.auxValue = max(p + valence * p, 0.0)
                    normIntrinsicProbabilities()
                    updateBehaviorNetNeuronLabels(behaviorNet)
                }
            }

            addButton("Reward") {
                learn(1.0)
                rewardNeuron.activation = 1.0
                punishNeuron.activation = 0.0
                SimbrainDesktop.workspace.iterateSuspend()
            }

            addButton("Punish") {
                learn(-1.0)
                rewardNeuron.activation = 0.0
                punishNeuron.activation = 1.0
                SimbrainDesktop.workspace.iterateSuspend()
            }

            addButton("Do nothing") {
                rewardNeuron.activation = 0.0
                punishNeuron.activation = 0.0
                SimbrainDesktop.workspace.iterateSuspend()
            }

        }
    }

}
